cmd/gke-certificates-controller/app/gke_approver.go

```diff
@@ -170,34 +189,138 @@ func hasExactUsages(csr *capi.CertificateSigningRequest, usages []capi.KeyUsage)
 	return true
 }
 
-var kubeletClientUsages = []capi.KeyUsage{
-	capi.UsageKeyEncipherment,
-	capi.UsageDigitalSignature,
-	capi.UsageClientAuth,
-}
+var (
+	kubeletClientUsages = []capi.KeyUsage{
+		capi.UsageKeyEncipherment,
+		capi.UsageDigitalSignature,
+		capi.UsageClientAuth,
+	}
+	kubeletServerUsages = []capi.KeyUsage{
+		capi.UsageKeyEncipherment,
+		capi.UsageDigitalSignature,
+		capi.UsageServerAuth,
+	}
+)
 
-func isNodeClientCert(csr *capi.CertificateSigningRequest, x509cr *x509.CertificateRequest) bool {
+func isNodeCert(opts approverOptions, csr *capi.CertificateSigningRequest, x509cr *x509.CertificateRequest) bool {
 	if !reflect.DeepEqual([]string{"system:nodes"}, x509cr.Subject.Organization) {
 		return false
 	}
 	if (len(x509cr.DNSNames) > 0) || (len(x509cr.EmailAddresses) > 0) || (len(x509cr.IPAddresses) > 0) {
 		return false
 	}
+	if !strings.HasPrefix(x509cr.Subject.CommonName, "system:node:") {
+		return false
+	}
+	return true
+}
+
+func isNodeClientCert(opts approverOptions, csr *capi.CertificateSigningRequest, x509cr *x509.CertificateRequest) bool {
+	if !isNodeCert(opts, csr, x509cr) {
+		return false
+	}
 	if !hasExactUsages(csr, kubeletClientUsages) {
 		return false
 	}
-	if !strings.HasPrefix(x509cr.Subject.CommonName, "system:node:") {
+	return true
+}
+
+func isSelfNodeClientCert(opts approverOptions, csr *capi.CertificateSigningRequest, x509cr *x509.CertificateRequest) bool {
+	if !isNodeClientCert(opts, csr, x509cr) {
+		return false
+	}
+	if csr.Spec.Username != x509cr.Subject.CommonName {
 		return false
 	}
 	return true
 }
 
-func isSelfNodeClientCert(csr *capi.CertificateSigningRequest, x509cr *x509.CertificateRequest) bool {
-	if !isNodeClientCert(csr, x509cr) {
+func isNodeServerCert(opts approverOptions, csr *capi.CertificateSigningRequest, x509cr *x509.CertificateRequest) bool {
+	if !isNodeCert(opts, csr, x509cr) {
+		return false
+	}
+	if !hasExactUsages(csr, kubeletServerUsages) {
 		return false
 	}
 	if csr.Spec.Username != x509cr.Subject.CommonName {
 		return false
 	}
 	return true
 }
+
+func validateNodeServerCert(opts approverOptions, csr *capi.CertificateSigningRequest, x509cr *x509.CertificateRequest) bool {
+	client := oauth2.NewClient(context.Background(), opts.tokenSource)
+
+	if err := validateNodeServerCertInner(client, opts, csr, x509cr); err != nil {
+		glog.Errorf("validating CSR %q: %v", csr.Name, err)
+		return false
+	}
+	return true
+}
+
+// Only check that DNS name and IP in SAN match an existing VM in the project.
+// Username was already checked against CN, so this CSR is coming from
+// authenticated kubelet.
+func validateNodeServerCertInner(client *http.Client, opts approverOptions, csr *capi.CertificateSigningRequest, x509cr *x509.CertificateRequest) error {
+	if len(x509cr.DNSNames) == 0 {
+		return errors.New("no SAN DNS names")
+	}
+	if len(x509cr.IPAddresses) == 0 {
```

Comment:
Also consider `if len(dnsNames) + len(idAddrs) == 0 { return err }`

Reply:
Also consider capping the number of SANs at something more than reasonable, e.g. 10 or 20.

Reply:
Also only allow dns and ip sans.